### ISSUE TITLE
nlp.pipe cannot pickle when using Transformer model

### ISSUE THREAD

**Issue by Minyall** [explosion/spaCy#13233]
Hello everyone. I'm having an issue running a transformer model using `.pipe` when I specify more than one process. Traceback suggests that something in the process cannot be pickled but I'm not sure why. The code below works with `en_core_web_md`, but switching to transformers causes the error below.

I suspect it may be a version incompatibility somewhere, so I have included dependency details from my Poetry project file, and the full lock file if helpful.

## How to reproduce the behaviour
```
import spacy

if __name__ == '__main__':

    N_WORKERS = -1

    texts = ['text'] * 100

    nlp = spacy.load('en_core_web_trf')
    for doc in nlp.pipe(texts,n_process=N_WORKERS):
        for e in doc.ents:
           print(e.text,e.label_)
```

```
Traceback (most recent call last):
  File "/Users/James/project/NER_tagging_script.py", line 9, in <module>
    for doc in nlp.pipe(texts,n_process=N_WORKERS):
  File "/Users/James/project/.venv/lib/python3.9/site-packages/spacy/language.py", line 1618, in pipe
    for doc in docs:
  File "/Users/James/project/.venv/lib/python3.9/site-packages/spacy/language.py", line 1684, in _multiprocessing_pipe
    proc.start()
  File "/opt/miniconda3/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/opt/miniconda3/lib/python3.9/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/opt/miniconda3/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/opt/miniconda3/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/opt/miniconda3/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/opt/miniconda3/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/opt/miniconda3/lib/python3.9/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
  File "stringsource", line 2, in curated_tokenizers._bbpe.ByteBPEProcessor.__reduce_cython__
TypeError: self._merges cannot be converted to a Python object for pickling
```

## Your Environment

- **spaCy version:** 3.7.2
- **Platform:** macOS-10.16-x86_64-i386-64bit
- **Python version:** 3.9.1
- **Pipelines:** en_core_web_md (3.7.1), en_core_web_trf (3.7.3)
* Environment Information:
```
[tool.poetry.dependencies]
python = "^3.8"
pandas = "^2.0.3"
jupyter = "^1.0.0"
plotly = "^5.15.0"
bertopic = "^0.15.0"
spacy = "^3.7.2"
en_core_web_md = {url = "https://github.com/explosion/spacy-models/releases/download/en_core_web_md-3.7.1/en_core_web_md-3.7.1.tar.gz"}
en_core_web_trf = {url = "https://github.com/explosion/spacy-models/releases/download/en_core_web_trf-3.7.3/en_core_web_trf-3.7.3.tar.gz"}
spacy-transformers = "^1.3.4"
```
Full env. lock file.
[poetry_lock.txt](https://github.com/explosion/spaCy/files/13919158/poetry_lock.txt)

**Comment by chrisgoddard** [comment]
Same issue here - Python 3.11 - digging into Cython code is a bit beyond me so not sure how do resolve.

**Comment by Minyall** [comment]
@chrisgoddard are you using Poetry for environment management too?

Just to note I tried making a clean env install with just spacy and its dependencies and still have the same behaviour.

I don't know if multiprocessing with transformers is not supported (or perhaps requires much higher spec hardware?) but I haven't found anything that suggests this.

**Comment by svlandeg** [comment]
Hi, thanks for the report! The 3.7.3 version of `en_core_web_trf` uses our relatively new [`curated-transformers`](https://github.com/explosion/curated-transformers/), so I wonder whether this is a new problem that crept in since then.

Any chance you could experiment with a slightly lower spaCy version (3.6) or below and a compatible trf model, to see whether this error still occurs?

**Comment by Minyall** [comment]
@svlandeg I think you're on to something. Made a clean env (specs below) and works as expected.

## Info about spaCy

- **spaCy version:** 3.6.1
- **Platform:** macOS-14.2.1-arm64-arm-64bit
- **Python version:** 3.8.11
- **Pipelines:** en_core_web_trf (3.6.1)

**Comment by svlandeg** [comment]
Thanks so much for checking! That confirms the hunch that the issue is in fact with `curated-transformers` or `spacy-curated-transformers`. We'll have to look into this.

**Comment by Minyall** [comment]
Glad it's not me missing something obvious! Ok, I can roll back to 3.6
until things get fixed, no bother.

**Comment by danieldk** [comment]
We have released Curated Tokenizers 0.0.9 which fixes this issue:
https://github.com/explosion/curated-tokenizers/releases/tag/v0.9.1

Unfortunately, for multiprocessing there is another issue:
https://github.com/explosion/spaCy/issues/13238

We are working on an update of the `spacy-curated-transformers` package which fixes that issue as well.

**Comment by svlandeg** [comment]
Update: this should be fixed after merging & releasing https://github.com/explosion/spacy-curated-transformers/pull/29 🤞

**Comment by danieldk** [comment]
We have released `spacy-curated-transformers` 0.2.2, which should fix the remaining issue with multiprocessing `pipe`.

Thanks again for reporting this issue! I'll close it for now, but feel free to re-open if you encounter further issues with pickling while using `Language.pipe`.
